pkg/endpoint/policy.go

```diff
@@ -163,81 +148,151 @@ func (e *Endpoint) setNextPolicyRevision(revision uint64) {
 // however, and it is possible that policy update succeeds for some endpoints,
 // while it fails for other endpoints.
 //
-// Returns:
-//   - err: any error in obtaining information for computing policy, or if
+// Failure may be due to any error in obtaining information for computing policy,
+// or if policy could not be generated given the current set of rules in the repository.
 //
-// policy could not be generated given the current set of rules in the
-// repository.
-// Must be called with endpoint mutex held.
-func (e *Endpoint) regeneratePolicy() (retErr error) {
-	var forceRegeneration bool
+// endpoint lock must NOT be held. This is because the ipcache needs to be able to
+// make progress while generating policy, and *that* needs the endpoint unlocked to call
+// ep.ApplyPolicyMapChanges. Specifically, computing policy may cause identity allocation
+// which requires ipcache progress.
+//
+// buildMutex MUST be held, and not released until setDesiredPolicy and
+// updateRealizedState have been called
+//
+// There are a few fields that depend on this exact configuration of locking:
+//   - ep.desiredPolicy: ep.mutex must be locked between writing this and committing to
+//     the policy maps, or else policy drops may occur
+//   - ep.policyRevision: ep.mutex and ep.buildMutex must be held to write to this, or else
+//     the policyRevision may go backwards from no-op bumps from policy ingestion
+//   - ep.selectorPolicy: this may be nulled if the endpoints identity changes; we must
+//     check for this when committing. ep.mutex must be held
+//   - ep.realizedRedirects: this is read by external callers as part of policy generation,
+//     so ep.mutex must not be required to read this. Instead, both ep.mutex and ep.buildMutex
+//     must be held to write to this (i.e. we are deep in regeneration)
+//
+// Returns a result that should be passed to setDesiredPolicy after the endpoint's
+// write lock has been acquired, or err if recomputing policy failed.
+func (e *Endpoint) regeneratePolicy() (*policyGenerateResult, error) {
+	var err error
+
+	// lock the endpoint, read our values, then unlock
+	err = e.rlockAlive()
+	if err != nil {
+		return nil, err
+	}
 
 	// No point in calculating policy if endpoint does not have an identity yet.
 	if e.SecurityIdentity == nil {
 		e.getLogger().Warn("Endpoint lacks identity, skipping policy calculation")
-		return nil
+		e.runlock()
+		return nil, nil
+	}
+
+	// Copy out some values we care about, then unlock
+	forcePolicyCompute := e.forcePolicyCompute
```

Comment:
Minor tricky part here: Previously, we held the lock, chose whether to force policy computation, then stored false (in `setDesiredPolicy()`), then released. However, I think that now that we fetch this outside the lock, it's possible for us to calculate policy for an old identity, then identity labels change (which would also set `e.forcePolicyCompute`), then call `setDesiredPolicy` with both (a) the old identity's policy and (b) determining that the next policy generation doesn't need to force policy computation. But actually it _does_ still need to force policy computation.

It feels like we almost need to clear the flag before calculating policy, then calculate policy, then iff (policy calculation failed && forcePolicyCompute was set), grab the lock and set the flag again. Plus remove the flag clearing in `setDesiredPolicy()`. Basically this way once we release the lock in `regeneratePolicy()` it's possible for some other caller to come in and ensure that the _next_ regeneration will also force policy calculation, taking into account any changes that may have occurred while policy calculation was not holding the lock.

Reply:
We already catch the first case (identity reset) by checking to see if `ep.identityRevision` changed during computation.

I think you're right; we should clear forcePolicyCompute in the first phase. Good idea, I'll do that.

Reply:
Ah yes I did see the identityRevision, didn't account for that in this line of reasoning. :+1: